### DISCUSSION TITLE
Help with issues with rebuilding moose

### DISCUSSION THREAD

**Discussion by Nabiel-Abuyazid** [idaholab/moose#24622] [Q&A Getting Started] [answered]
Hi,

I made some changes to the codebase accidentally. So instead of trying to find where the issue is specifically, I decided to rebuild moose. However, when doing so, this spits out:

/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/libtool: No such file or directory
/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/libtool: No such file or directory
/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
find: /Users/nabzforcemac/projects/moose/framework/contrib/wasp/install/lib: No such file or directory
Rebuilding symlinks in /Users/nabzforcemac/projects/moose/framework/build/header_symlinks
Rebuilding symlinks in /Users/nabzforcemac/projects/moose/test/build/header_symlinks
Creating Unity Directory /Users/nabzforcemac/projects/moose/framework/build/unity_src
Checking if header needs updating: /Users/nabzforcemac/projects/moose/framework/include/base/MooseRevision.h...
Copying default MOOSE configuration to: /Users/nabzforcemac/projects/moose/framework/include/base/MooseConfig.h...
/bin/sh: line 0: [: -ne: unary operator expected
/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
Compiling C++ (in opt mode) /Users/nabzforcemac/projects/moose/framework/contrib/pcre/src/pcrecpp.cc...
/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
Compiling C++ (in opt mode) /Users/nabzforcemac/projects/moose/framework/contrib/pcre/src/pcre_stringpiece.cc...
/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/libtool: No such file or directory
/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
make: *** [/Users/nabzforcemac/projects/moose/framework/contrib/pcre/src/pcrecpp..opt.lo] Error 127
make: *** Waiting for unfinished jobs....
/bin/sh: /Users/nabzforcemac/projects/moose/libmesh/installed/libtool: No such file or directory
make: *** [/Users/nabzforcemac/projects/moose/framework/contrib/pcre/src/pcre_stringpiece..opt.lo] Error 127

**Comment by milljm** [comment]
I can't tell if you have activate the necessary Conda environment or not. But this usually occurs when one forgets to:

```bash
mamba activate moose
```
before running make.

This is a little weird though:
```pre
/bin/sh: line 0: [: -ne: unary operator expected
```
Not sure where that's coming from.

**Reply by Nabiel-Abuyazid** [reply]
Thanks Jason that helped solve the issue, but now I am running into these errors:

/Users/nabzforcemac/projects/moose/framework/src/outputs/PetscOutput.C:151:76: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'                                                                                                        
    PetscErrorCode ierr = SNESMonitorSet(snes, petscNonlinearOutput, this, LIBMESH_PETSC_NULLPTR);                                                                                                                                                
                                                                           ^                                                                                                                                                                      
/Users/nabzforcemac/projects/moose/framework/src/outputs/PetscOutput.C:158:71: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'                                                                                                        
    PetscErrorCode ierr = KSPMonitorSet(ksp, petscLinearOutput, this, LIBMESH_PETSC_NULLPTR);                                                                                                                                                     
                                                                      ^                                                                                                                                                                           
2 errors generated.                                                                                                                                                                                                                               
make: *** [/Users/nabzforcemac/projects/moose/framework/build.mk:149: /Users/nabzforcemac/projects/moose/framework/build/unity_src/outputs_Unity.arm64-apple-darwin20.0.0.opt.lo] Error 1                                                         
make: *** Waiting for unfinished jobs....

**Reply by milljm** [reply]
It looks like there was an update involved (you updated the moose repo, or you updated Conda). When one occurs, so must the other. I am not sure which was done, so we will do them both:

```bash
mamba activate moose
cd ~/projects/moose
git pull
mamba update --all
```

Then, it is necessary to clean up any stale/old build files lying around in the moose repo:

```bash
git clean -xfd
```
Note: this will delete any untracked files. So if you are making changes to moose, be sure to save those changes before running `git clean`.

Once this is all complete, you should be able to run `make` again.

**Reply by Nabiel-Abuyazid** [reply]
I have followed your instructions, but some issues still persist:

/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:529:20: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
    dmm->_splits = LIBMESH_PETSC_NULLPTR;
                   ^
/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:534:23: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
    dmm->_splitlocs = LIBMESH_PETSC_NULLPTR;
                      ^
/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:541:16: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
    info._dm = LIBMESH_PETSC_NULLPTR;
               ^
/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:542:24: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
    info._rembedding = LIBMESH_PETSC_NULLPTR;
                       ^
/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:1065:20: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
    *outerislist = LIBMESH_PETSC_NULLPTR; /* FIX: allow mesh-based overlap. */
                   ^
/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:1618:19: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
    dmm->_sides = LIBMESH_PETSC_NULLPTR;
                  ^
/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:1634:21: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
    dmm->_unsides = LIBMESH_PETSC_NULLPTR;
                    ^
/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:1715:18: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
    dmm->_vars = LIBMESH_PETSC_NULLPTR;
                 ^
/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:1772:20: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
    dmm->_blocks = LIBMESH_PETSC_NULLPTR;
                   ^
Compiling C++ (in opt mode) /Users/nabzforcemac/projects/moose/framework/src/utils/SlepcSupport.C...
/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:1977:66: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
      opt.c_str(), help.c_str(), "DMMooseSetVars", vars, &nvars, LIBMESH_PETSC_NULLPTR);
                                                                 ^
/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:2005:72: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
      opt.c_str(), help.c_str(), "DMMooseSetBlocks", blocks, &nblocks, LIBMESH_PETSC_NULLPTR);
                                                                       ^
/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:2029:69: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
      opt.c_str(), help.c_str(), "DMMooseSetSides", sides, &nsides, LIBMESH_PETSC_NULLPTR);
                                                                    ^
/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:2046:71: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
      opt.c_str(), help.c_str(), "DMMooseSetUnSides", sides, &nsides, LIBMESH_PETSC_NULLPTR);
                                                                      ^
/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:2083:26: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
                         LIBMESH_PETSC_NULLPTR);
                         ^
Compiling C++ (in opt mode) /Users/nabzforcemac/projects/moose/framework/src/utils/KDTree.C...
/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:2105:38: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
                                     LIBMESH_PETSC_NULLPTR);
                                     ^
/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:2132:31: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
                              LIBMESH_PETSC_NULLPTR);
                              ^
/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:2152:29: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
                            LIBMESH_PETSC_NULLPTR);
                            ^
/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:2171:26: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
                         LIBMESH_PETSC_NULLPTR);
                         ^
/Users/nabzforcemac/projects/moose/framework/src/utils/PetscDMMoose.C:2193:38: error: use of undeclared identifier 'LIBMESH_PETSC_NULLPTR'
                                     LIBMESH_PETSC_NULLPTR);
                                     ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.

**Reply by milljm** [reply]
Hmm, odd! Can you perform the following:

```bash
cd ~/projects/moose
scripts/versioner.py libmesh --yaml | grep install
mamba list | grep moose
```

**Reply by Nabiel-Abuyazid** [reply]
Done. This is what the output is:

ImportError: dlopen(/Users/nabzforcemac/mambaforge3/lib/python3.10/site-packages/libmambapy/bindings.cpython-310-darwin.so, 0x0002): Library not loaded: @rpath/libarchive.13.dylib
  Referenced from: <96FF200C-F283-35DD-A021-9DED6415FDD2> /Users/nabzforcemac/mambaforge3/lib/libmamba.2.0.0.dylib
  Reason: tried: '/Users/nabzforcemac/mambaforge3/lib/libarchive.13.dylib' (no such file), '/Users/nabzforcemac/mambaforge3/lib/python3.10/site-packages/libmambapy/../../../libarchive.13.dylib' (no such file), '/Users/nabzforcemac/mambaforge3/lib/python3.10/site-packages/libmambapy/../../../libarchive.13.dylib' (no such file), '/Users/nabzforcemac/mambaforge3/bin/../lib/libarchive.13.dylib' (no such file), '/Users/nabzforcemac/mambaforge3/bin/../lib/libarchive.13.dylib' (no such file), '/usr/local/lib/libarchive.13.dylib' (no such file), '/usr/lib/libarchive.13.dylib' (no such file, not in dyld cache)

**Reply by milljm** [reply]
There might be something wrong with your PATHs  being created in your startup profile. To be sure, can you open a new terminal, and run the above commands again. But this time, include everything. Including the command I am asking you to run. Example:

```bash
Last login: Wed Jun  7 11:32:47 on ttys013
❯ cd ~/projects/moose
scripts/versioner.py libmesh --yaml | grep install
❯ mamba list | grep moose
  install: moose-libmesh=2023.04.19=build_1
❯
```

**Reply by Nabiel-Abuyazid** [reply]
Here it is:

Last login: Wed Jun  7 13:45:39 on ttys006
-bash: export: `=': not a valid identifier

The default interactive shell is now zsh.
To update your account to use zsh, please run `chsh -s /bin/zsh`.
For more details, please visit https://support.apple.com/kb/HT208050.
(base) Nabiels-MacBook-Pro:~ nabzforcemac$ cd ~/projects/moose
(base) Nabiels-MacBook-Pro:moose nabzforcemac$ scripts/versioner.py libmesh --yaml | grep install
Traceback (most recent call last):
  File "/Users/nabzforcemac/projects/moose/scripts/versioner.py", line 25, in <module>
    import yaml
ModuleNotFoundError: No module named 'yaml'
(base) Nabiels-MacBook-Pro:moose nabzforcemac$ mamba list | grep moose
-bash: __add_sys_prefix_to_path: command not found
Traceback (most recent call last):
  File "/Users/nabzforcemac/mambaforge3/bin/mamba", line 7, in <module>
    from mamba.mamba import main
  File "/Users/nabzforcemac/mambaforge3/lib/python3.10/site-packages/mamba/mamba.py", line 49, in <module>
    import libmambapy as api
  File "/Users/nabzforcemac/mambaforge3/lib/python3.10/site-packages/libmambapy/__init__.py", line 7, in <module>
    raise e
  File "/Users/nabzforcemac/mambaforge3/lib/python3.10/site-packages/libmambapy/__init__.py", line 4, in <module>
    from libmambapy.bindings import *  # noqa: F401,F403
ImportError: dlopen(/Users/nabzforcemac/mambaforge3/lib/python3.10/site-packages/libmambapy/bindings.cpython-310-darwin.so, 0x0002): Library not loaded: @rpath/libarchive.13.dylib
  Referenced from: <96FF200C-F283-35DD-A021-9DED6415FDD2> /Users/nabzforcemac/mambaforge3/lib/libmamba.2.0.0.dylib
  Reason: tried: '/Users/nabzforcemac/mambaforge3/lib/libarchive.13.dylib' (no such file), '/Users/nabzforcemac/mambaforge3/lib/python3.10/site-packages/libmambapy/../../../libarchive.13.dylib' (no such file), '/Users/nabzforcemac/mambaforge3/lib/python3.10/site-packages/libmambapy/../../../libarchive.13.dylib' (no such file), '/Users/nabzforcemac/mambaforge3/bin/../lib/libarchive.13.dylib' (no such file), '/Users/nabzforcemac/mambaforge3/bin/../lib/libarchive.13.dylib' (no such file), '/usr/local/lib/libarchive.13.dylib' (no such file), '/usr/lib/libarchive.13.dylib' (no such file, not in dyld cache)
(base) Nabiels-MacBook-Pro:moose nabzforcemac$

**Reply by milljm** [reply]
Well there is something wrong with or how mamba is being initialized. Due to the strange operator error that bash is mentioning, it might be due to something at odds in your `~/.bash_profile`.

I am guessing you have the initialization routines at the top of the `~/.bash_profile`, and later on, something is messing around with paths variables which are affecting Mamba somehow...

Can you show us your bash profile? Might be able to help:

```bash
cat ~/.bash_profile
```

**Reply by Nabiel-Abuyazid** [reply]
Ah I see. Okay, here it is:


```
# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/Users/nabzforcemac/mambaforge3/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/Users/nabzforcemac/mambaforge3/etc/profile.d/conda.sh" ]; then
        . "/Users/nabzforcemac/mambaforge3/etc/profile.d/conda.sh"
    else
        export PATH="/Users/nabzforcemac/mambaforge3/bin:$PATH"
    fi
fi
unset __conda_setup

if [ -f "/Users/nabzforcemac/mambaforge3/etc/profile.d/mamba.sh" ]; then
    . "/Users/nabzforcemac/mambaforge3/etc/profile.d/mamba.sh"
fi
# <<< conda initialize <<<

export = PATH=$PATH:$HOME/projects/moose/python/peacock/
```

**Reply by milljm** [reply, marked as answer]
While that may solve the operator error... I think that that PATH= line would not break Mamba...

You might need to reinstall Mamba. If you've only been using Conda for MOOSE based stuff, it should be safe to delete, and re-install...

```bash
rm -rf /Users/nabzforcemac/mambaforge3
```

- Apple Silicon Mac:

  ```bash
  curl -L -O https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-MacOSX-arm64.sh
  sh Mambaforge-MacOSX-arm64.sh -b -p /Users/nabzforcemac/mambaforge3
  ```

- Intel Mac:

  ```bash
  curl -L -O https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-MacOSX-x86_64.sh
  sh Mambaforge-MacOSX-x86_64.sh -b -p /Users/nabzforcemac/mambaforge3
  ```

Exit your terminal, and re-open it... see if you can run `mamba` commands without those errors:

```bash
mamba list
```
If the above works without those crazy errors, you'll need to re-create the 'moose' environment:

```bash
mamba create -n moose moose-dev
mamba activate moose
```

and hopefully, you should be able to build once more.

**Reply by ykvishal** [reply]
@Nabiel-Abuyazid 
I think that you need to do first 
`mamba activate moose`

then run 
`scripts/versioner.py libmesh --yaml | grep install`
`mamba list | grep moose`

also your `~/.bash_profile` has a wrong path for peacock `export PATH=pwd:$PATH`
it should be `export PATH=/Users/nabzforcemac/projects/moose/python/peacock/:$PATH`

**Reply by Nabiel-Abuyazid** [reply]
Thank you for your help Jason!

I have another question that I was hoping you would be able to answer:

I am trying to create a custom kernel similar to CoefTimeDerivative, but instead of a coefficient, it is a material property. I have named it MatTimeDerivative. But during the make step I seem to be getting errors. Truthfully, I am not familiar with C++ and all of the classes that are needed to properly incorporate a material property. I have posted it as #24485 

If you could take a look at it, I would greatly appreciate it!

**Reply by milljm** [reply]
I can't help you there... I am not a finite element researcher, nor a C++ developer... You'll have better luck in your new discussion #24485 when someone else from our team contacts you.
